lib/heap_sort.rb

```diff
@@ -1,8 +1,72 @@
-
+require_relative "min_heap"
 
 # This method uses a heap to sort an array.
-# Time Complexity:  ?
-# Space Complexity: ?
-def heap_sort(list)
-  raise NotImplementedError, "Method not implemented yet..."
+# Time Complexity:  O(n logn)
+# Space Complexity: O(n), made a new [] and new MinHeap instance
+def heapsort(list)
+  heap = MinHeap.new()
+  
+  # add each element onto MinHeap
+  list.each do |element|
+    heap.add(element)
+  end
+  
+  # start plucking out the roots from the MinHeap
+  sortedList = []
+  until heap.empty?
+    sortedList << heap.remove()
+  end
+  
+  return sortedList
+end
+
+
+
+
+
+
+################## O(1) space complexity with sorting in place ##################
+### Keeping print statements in there on purpose, makes it easier to understand ###
+
+# see heapsort_inplace.jpg for how the orig list and new MinHeap are sharing the same space
+# basically the heap grows in the list space as the unsorted shares drop, and vice versa
+def heapsort_inPlace(list)
```

Comment:
This is in place but because you're deleting each element this is an O(n^2 * log n) sorting algorithm.